File: AES-CBC.js
const CryptoJS = require("crypto-js");

function aesCbcEncrypt(key, iv, plaintext) {
  const encrypted = CryptoJS.AES.encrypt(plaintext, key, {
    iv: iv,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  return encrypted.toString();
}

function aesCbcDecrypt(key, iv, ciphertext) {
  const decrypted = CryptoJS.AES.decrypt(ciphertext, key, {
    iv: iv,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  return decrypted.toString(CryptoJS.enc.Utf8);
}

const key = CryptoJS.enc.Hex.parse("00112233445566778899aabbccddeeff"); // 256-bit key
const iv = CryptoJS.enc.Hex.parse("aabbccddeeff00112233445566778899"); // 128-bit IV
const plaintext = "This is a secret message";

console.time("Encryption Time");
const ciphertext = aesCbcEncrypt(key, iv, plaintext);
console.timeEnd("Encryption Time");

console.time("Decryption Time");
const decryptedText = aesCbcDecrypt(key, iv, ciphertext);
console.timeEnd("Decryption Time");

console.log("Ciphertext:", ciphertext);
console.log("Decrypted Text:", decryptedText);

File: AES-CBC.py
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
import base64
import time

def pad(data):
    block_size = 16
    padding = block_size - len(data) % block_size
    return data + bytes([padding] * padding)

def unpad(data):
    padding = data[-1]
    return data[:-padding]

def aes_cbc_encrypt(key, plaintext):
    iv = get_random_bytes(16)
    cipher = AES.new(key, AES.MODE_CBC, iv)
    plaintext = pad(plaintext)  # Apply PKCS#7 padding
    ciphertext = cipher.encrypt(plaintext)
    return base64.b64encode(iv + ciphertext)

def aes_cbc_decrypt(key, ciphertext):
    ciphertext = base64.b64decode(ciphertext)
    iv, ciphertext = ciphertext[:16], ciphertext[16:]
    cipher = AES.new(key, AES.MODE_CBC, iv)
    plaintext = cipher.decrypt(ciphertext)
    plaintext = unpad(plaintext)  # Remove PKCS#7 padding
    return plaintext

key = get_random_bytes(16)
plaintext = b'This is a secret message.'

# Do thoi gian danh cho ma hoa
start_time = time.time()
ciphertext = aes_cbc_encrypt(key, plaintext)
end_time = time.time()
print("Encryption Time:", (end_time - start_time)*1000 , "ms")

# Do thoi gian danh cho giai ma
start_time = time.time()
decrypted_text = aes_cbc_decrypt(key, ciphertext)
end_time = time.time()
print("Decryption Time:", (end_time - start_time)*1000 , "ms")

print("Ciphertext:", ciphertext)
print("Decrypted Text:", decrypted_text.decode('utf-8'))
